---
title: "evolutionary_analysis"
author: "Sahil Shah"
date: "7/2/2019"
output: pdf_document
---

```{r}

library(reshape2)
library(gridExtra)
library(gtools)
#Set working directory to the directory in which the results are contained in 
setwd("../../data/targets/")
#setwd("../results/2020_4_30/paper_data8_nf1.0_rep8_nmut5/")
#setwd("/home/sahil/")

#Change grant_data name to the target file used
#target_data <- read.table("expression_pattern_757.tsv", header=TRUE)
#target_data <- read.table("test.tsv", header=TRUE)
target_data <- read.table("paper_data1.tsv", header=TRUE)
target_data <- filter(target_data, species == "protein1" | species == "protein2" | species == "protein3")
target_line_plot <-  ggplot(target_data, aes(fill=species, color=species, size=species, x=time, y=transcript)) + geom_line(stat="identity", size=2) + ggtitle("Target Genome") + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(color = "black", size=1), axis.ticks = element_line(colour = "black", size = 1), axis.text=element_text(color="black",size=30,face="bold"), axis.title=element_text(size=30,face="bold"), plot.title=element_text(size = 25, face = "bold")) + scale_color_manual(values=c("#0072B2", "#E69F00", "#CC79A7")) + ylab('transcript abundances')

ggsave("../figures/paper_data1.png", target_line_plot, width=10, height=10, limitsize=FALSE)
#ggsave("test.png", target_line_plot, width=10, height=10, limitsize=FALSE)

```
Saves sum of squares data
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(cowplot)
library(dplyr)
library(gridExtra)
library(ggplot2)
library(grid)
#Set working directory to the directory in which the results are contained in
setwd("../../results/2020_4_30/paper_data13_nf1.0_rep8_nmut5/")

all_rmse_data <- read.table("rmse_data.tsv", header=TRUE)
rmse_data_acc <- all_rmse_data %>% filter(Accepted=='yes')

#rmse_line_plot <- ggplot(rmse_data_acc, aes(x=Iteration, y=Sum_of_Squares, color='red')) + geom_point(aes(color='red'), stat="identity") + geom_line(aes(color='red'), size= 1, stat="identity") + geom_line(data=all_rmse_data, aes(x=Iteration, y=Sum_of_Squares), color='black') + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
#panel.background = element_blank(), axis.line = element_line(color = "black")) + xlab("Sum of Squares") + ylab("Iteration")
rmse_line_plot <- ggplot(rmse_data_acc, aes(x=Iteration, y=Sum_of_Squares)) + geom_point(stat="identity") + geom_line(size= 1, stat="identity") + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(color = "black", size=1), axis.ticks = element_line(colour = "black", size = 1), axis.text=element_text(size=30,face="bold"), axis.title=element_text(size=30,face="bold")) + xlab("Generation") + ylab("Sum of Squares")
rmse_plots <- grid.arrange(rmse_line_plot, nrow=1, top=textGrob("Fitness over Time", gp=gpar(fontsize=30, fontface="bold")))
#Change name of rmse_graphs to a unique identifier
ggsave("../../../data/figures/rmse_graphs_paper_data13_rep8_nf1.png", rmse_plots, width=20, height=10)
```

Evolutionary Analysis of Mutations

```{r pressure, echo=FALSE}
library(reshape2)
library(gridExtra)
library(gtools)
#Set working directory to the directory in which the results are contained in 
setwd("../../results/2020_8_7/paper_data1_rep6_nmut10/")

plots_list <- list()
best_files <- list.files(pattern="expression_pattern_")
best_files <- mixedsort(sort(best_files))
plots_list <- list()
# #Change grant_data name to the target file used
# target_data <- read.table("../../../data/targets/paper_data1.tsv", header=TRUE)
# target_data <- filter(target_data, species == "protein1" | species == "protein2" | species == "protein3")
# target_line_plot <-  ggplot(target_data, aes(fill=species, color=species, size=species, x=time, y=transcript)) + geom_line(stat="identity", size=2) + ggtitle("Target Genome Function") + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(color = "black", size=1), axis.ticks = element_line(colour = "black", size = 1), axis.text=element_text(color="black",size=30,face="bold"), axis.title=element_text(size=30,face="bold"), plot.title=element_text(size = 25, face = "bold")) + scale_color_manual(values=c("#0072B2", "#E69F00", "#CC79A7")) + panel_border()
#   plots_list[[1]] <- target_line_plot
#   #Determines if there are too many accepted mutations and, if so, only display the last 20 that were accepted
#   starting_point <- 1
#   if (length(best_files) > 40){
#     starting_point <- length(best_files) - 40
#   }
#   plots_start <- 1
#   #Creates plots for each accepted mutation
#   for(i in starting_point:length(best_files)){
#     plots_start <- plots_start + 1
#     best <- "best"
#     evolution_title <- "Genome Evolution "
#     plot_title <- paste(evolution_title, i, sep="")
#     file_name <- assign(paste(best, i, sep=""), read.table(best_files[i], fill=TRUE, header=TRUE))
#     file_name <- filter(file_name, species == "protein1" | species == "protein2" | species == "protein3")
#     plot <- ggplot(file_name, aes(fill=species, color=species, size=species, x=time, y=transcript)) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(color = "black", size=1), axis.ticks = element_line(colour = "black", size = 1), axis.text=element_text(color="black",size=30,face="bold"), axis.title=element_text(size=30,face="bold"), plot.title=element_text(size = 25, face = "bold")) + geom_line(stat="identity", size=2) + ggtitle(plot_title) + scale_color_manual(values=c("#0072B2", "#E69F00", "#CC79A7"))
#     plots_list[[plots_start]] <- plot
#   }
  
# #Outputs file with the generated plots
# evo_plot <- grid.arrange(grobs=plots_list, nrow=ceiling(length(plots_list) / 4), ncol=4, top=textGrob("Evolution of Gene Expression Pattern", gp=gpar(fontsize=30, fontface="bold")))
# #Change name of saved graphs file to a unique identifier
# ggsave("../../../data/figures/paper_data139_rep0_nf1000.png", evo_plot, width=30, height=30, limitsize=FALSE)
# # 
# rmse_data <- read.table("./final/rmse_data.tsv", header=TRUE)
# rmse_data <- rmse_data %>% filter(Accepted=='yes')
# min_rmse <- rmse_data$Iteration[rmse_data$NRMSE == min(rmse_data$NRMSE)]
# best_file <- list.files(pattern=paste0("expression_pattern_", min_rmse, ".tsv"))
# plot_data <- read.table("./final/expression_pattern_best.tsv", header=TRUE)
# plot_data <- filter(plot_data, species == "protein1" | species == "protein2" | species == "protein3")
# line_plot <-  ggplot(plot_data, aes(fill=species, color=species, size=species, x=time, y=transcript)) + geom_line(stat="identity", size=2) + ggtitle("Lowest target genome") + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(color = "black", size=1), axis.ticks = element_line(colour = "black", size = 1), axis.text=element_text(color="black",size=30,face="bold"), axis.title=element_text(size=30,face="bold"), plot.title=element_text(size = 25, face = "bold")) + scale_color_manual(values=c("#0072B2", "#E69F00", "#CC79A7")) + ylab('Transcript abundances')
# ggsave("../../../data/figures/paper_data64_rep6_lowest.png", line_plot, width=10, height=10, limitsize=FALSE)


#Filter sum of squares data
all_rmse_data <- read.table("./final/rmse_data.tsv", header=TRUE)
rmse_data_acc <- all_rmse_data %>% filter(Accepted=='yes')

#Create sum of squares plot 
rmse_line_plot <- ggplot(rmse_data_acc, aes(x=Iteration, y=NRMSE)) + 
  geom_point(stat='identity', color='#D55E00', size=3) + 
  geom_line(size=2, stat='identity', color='#D55E00') + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        plot.margin = unit(c(0.5,1,1,1), "cm"), 
        axis.text.x = element_text(size=39), 
        axis.text.y = element_text(size=39), 
        axis.line = element_line(size=1.3), 
        axis.ticks = element_line(size=1.3), 
        axis.ticks.length = unit(0.3, "cm"),
        axis.title = element_text(size=39),
        legend.position = "none") +
  xlab("Generation") + 
  ylab("Root mean square error") + 
  panel_border() +
  scale_y_continuous(limits = c(0, 3), expand = c(0, 0))

rmse_line_plot <- arrangeGrob(rmse_line_plot, top=textGrob("B", x = unit(0, "npc"), y=unit(1, "npc"), just=c("left","top"), gp=gpar(col="black", font="bold", fontsize=60, fontfamily="Arial")))
ggsave("../../../data/figures/paper_data1_rep6_rmse.png", rmse_line_plot, width=20, height=10, limitsize=FALSE)

```

```{r}
setwd("../../results/2020_7_8/paper_data1_rep202_nmut10/")
rmse_data <- read.table("./final/rmse_data.tsv", header=TRUE)
rmse_data <- rmse_data %>% filter(Accepted=='yes')
min_rmse <- rmse_data$Iteration[rmse_data$NRMSE == min(rmse_data$NRMSE)]
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
    