---
title: "evolutionary_analysis"
author: "Sahil Shah"
date: "7/2/2019"
output: pdf_document
---

Saves sum of squares data
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(cowplot)
library(dplyr)
library(gridExtra)
library(ggplot2)

setwd("/home/sahil/pinetree-toys/results/2019_10_14/random_data5_nf10_rep1/")

all_sos_data <- read.table("sos_data.tsv", header=TRUE)
sos_data_acc <- all_sos_data %>% filter(Accepted=='yes')

sos_line_plot <- ggplot(sos_data_acc, aes(x=Iteration, y=Sum_of_Squares, color='red')) + geom_point(aes(color='red'), stat="identity") + geom_line(aes(color='red'), size= 1, stat="identity") + geom_line(data=all_sos_data, aes(x=Iteration, y=Sum_of_Squares), color='black') + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(color = "black")) + xlab("Sum of Squares") + ylab("Iteration")
sos_plots <- grid.arrange(sos_line_plot, nrow=1, top="Sum of Squares Data")
ggsave("/home/sahil/pinetree-toys/results/2019_10_14/random_data5_nf10_rep1/sos_graphs.pdf", sos_plots, width=20, height=10)
```

Evolutionary Analysis of Mutations

```{r pressure, echo=FALSE}
library(reshape2)
library(gridExtra)
library(gtools)

setwd("/home/sahil/pinetree-toys/data/")

plots_list <- list()

best_files <- list.files("/home/sahil/pinetree-toys/data/", pattern="three_genes_replicated_")
best_files <- mixedsort(sort(best_files))
plots_list <- list()
  target_data <- read.table("/home/sahil/pinetree-toys/data/three_genes_replicated_1439.tsv", header=TRUE)
target_data <- filter(target_data, species == "protein1" | species == "protein2" | species == "protein3")
target_line_plot <-  ggplot(target_data, aes(fill=species, color=species, x=time, y=transcript)) + geom_line(stat="identity") + ggtitle("Target Genome Function") + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"))

for(i in 1:length(best_files)){
  best <- "best"
  best_plot <- "best_plot"
  evolution_title <- "Genome Evolution "
  plot_title <- paste(evolution_title, i, sep="")
  file_name <- assign(paste(best, i, sep=""), read.table(best_files[i], fill=TRUE, header=TRUE))
  file_name <- filter(file_name, species == "protein1" | species == "protein2" | species == "protein3")
  plot_name <- assign(paste(best_plot, i, sep=""), ggplot(file_name, aes(fill=species, color=species, x=time, y=transcript)) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(colour = "black")) + geom_line(stat="identity") + ggtitle(plot_title))
  plot <- ggplot(file_name, aes(fill=species, color=species, x=time, y=transcript)) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(colour = "black")) + geom_line(stat="identity") + ggtitle(plot_title)
  plots_list[[i]] <- plot

}

#evo_plot <- grid.arrange(plots_list, nrow=ceiling(length(best_files) / 2), ncol=ceiling(length(best_files) / 2), top="Evolutionary Pattern of Gene Transcripts") 
evo_plot <- grid.arrange(target_line_plot, nrow=1, ncol=1, top="Evolutionary Pattern of Gene Transcripts")
ggsave("/home/sahil/pinetree-toys/data/figures/three_genes_replicated_1439.pdf", evo_plot, width=30, height=30, limitsize=FALSE)

```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
