---
title: "Testing Parameters"
author: "Sahil Shah"
date: "4/22/2020"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(cowplot)
library(dplyr)
library(gridExtra)
library(ggplot2)
library(grid)
library(reshape2)
library(gtools)
library(png)
```
Function that returns the mean of tsv files are time point 239 seconds
```{r}

get_limits <- function(all_files){
  ratios <- c()
  for (f in 1:length(all_files)){
    protein1 <- read.table(all_files[f], header=TRUE, fill=TRUE) %>% filter(time==239, species=='protein1') %>% select('transcript')
    protein2 <- read.table(all_files[f], header=TRUE, fill=TRUE) %>% filter(time==239, species=='protein2') %>% select('transcript')
    if(nrow(protein2) != 0){
      ratios[f] <- protein2$transcript/protein1$transcript
    }else{
      ratios[f] <- 0
    }
  }
  return(sd(ratios)/sqrt(length(ratios)))
}

```
Plot promoter parameters
```{r}

setwd('../../data/parameter_testing/promoter/average/')

#List promoter files
prom_files <- list.files(pattern="promoter_average_test_")
prom_files <- mixedsort(sort(prom_files))

#Set x-axis scale
prom_x_data <- c(10e5, 25e5, 50e5, 75e5, 10e6, 25e6, 50e6, 75e6, 10e7, 25e7, 50e7, 75e7, 10e8)
prom_x_names <- c("10e5", "25e5", "50e5", "75e5", "10e6", "25e6", "50e6", "75e6", "10e7", "25e7", "50e7", "75e7", "10e8")
prom_y_data <- c()

#Get transcript abundace at time point 239 seconds
for (i in 1:length(prom_files)){
  p1_data <- read.table(prom_files[i], header=TRUE) %>% filter(time==239, species=='protein1') %>% select('transcript')
  p2_data <- read.table(prom_files[i], header=TRUE) %>% filter(time==239, species=='protein2') %>% select('transcript')
  if (nrow(p2_data) != 0){
    prom_y_data[i] <- p2_data$transcript/p1_data$transcript
  }else{
    prom_y_data[i] <- 0
  }
}

#Create dataframe for ggplot
prom_data <- data.frame(cbind(prom_x_data, prom_y_data))
colnames(prom_data) <- c('strengths','values')

#Enters working directory where the averages were calculated from
setwd('../')

#Gets averages of each numerical value for error bars
prom_means <- c()
for(p in 1:length(prom_x_data)){
  prom <- list.files(pattern=paste("promoter_test_", prom_x_names[p], sep=''))
  prom <- mixedsort(sort(prom))
  prom_means[p] <- get_limits(prom)
}

#Create plots
prom_line_plot <- ggplot(prom_data, aes(x=strengths, y=values)) + 
  geom_point(stat='identity', color='#009E73', size=5) + 
  geom_line(size=4, stat='identity', color='#009E73') + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        plot.margin = unit(c(2,2,2,2), "cm"), 
        axis.text.x = element_text(size=80), 
        axis.text.y = element_text(size=80), 
        axis.line = element_line(size=1.3), 
        axis.ticks = element_line(size=1.3), 
        axis.ticks.length = unit(0.5, "cm"),
        axis.title.x = element_text(size=80), 
        axis.title.y = element_text(size=80)) +
  xlab("Promoter strength (log scale)") + 
  ylab("Gene Y protein abundance/\nGene X protein abundance") + 
  panel_border() +
  geom_errorbar(aes(ymin=prom_y_data-prom_means, ymax=prom_y_data+prom_means), width=0.05, size=2) +
  scale_x_log10() +
  geom_vline(xintercept = 1e7, linetype="dashed", color = "dark grey", size=1.3)

#Log scaled plot
prom_log_plot <- ggplot(prom_data, aes(x=strengths, y=values)) + 
  geom_point(stat='identity', color='#009E73', size=5) + 
  geom_line(size=4, stat='identity', color='#009E73') + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        plot.margin = unit(c(2,2,2,2), "cm"), 
        axis.text.x = element_text(size=80), 
        axis.text.y = element_text(size=80), 
        axis.line = element_line(size=1.3), 
        axis.ticks = element_line(size=1.3), 
        axis.ticks.length = unit(0.5, "cm"),
        axis.title.x = element_text(size=80), 
        axis.title.y = element_text(size=80)) +
  xlab("Promoter strength (log scale)") + 
  ylab("Gene Y protein abundance/\nGene X protein abundance (log scale)") + 
  panel_border() +
  geom_errorbar(aes(ymin=prom_y_data-prom_means, ymax=prom_y_data+prom_means), width=0.05, size=2) +
  scale_y_log10() +
  scale_x_log10() +
  geom_vline(xintercept = 1e7, linetype="dashed", color = "dark grey", size=1.3)

```
Plot terminator parameters

```{r, echo=FALSE}

setwd('../../data/parameter_testing/terminator/average/')

#List terminator files
term_files <- list.files(pattern="terminator_average_test_")
term_files <- mixedsort(sort(term_files))

#Set x-axis scale
term_x_data <- c(0.0, 0.05, 0.1,
            0.15, 0.2,
            0.25, 0.3,
            0.35, 0.4,
            0.45, 0.5,
            0.55, 0.6,
            0.65, 0.7,
            0.75, 0.8,
            0.85, 0.9,
            0.95, 1.0)
            
term_y_data <- c()

#Get transcript abundaces at time point 239 seconds and calculate ratio between protein2 and protein1 abundances
for (i in 1:length(term_files)){
  p1_data <- read.table(term_files[i], header=TRUE) %>% filter(time==239, species=='protein1') %>% select('transcript')
  p2_data <- read.table(term_files[i], header=TRUE) %>% filter(time==239, species=='protein2') %>% select('transcript')
  if (nrow(p2_data) != 0){
    term_y_data[i] <- p2_data$transcript/p1_data$transcript
  }else{
    term_y_data[i] <- 0
  }
}

#Create dataframe needed for ggplot
term_data <- data.frame(cbind(term_x_data, term_y_data))
colnames(term_data) <- c('strengths','values')

#Enters working directory where the averages were calculated from
setwd('../')

#Gets averages of each numerical value for error bars
term_means <- c()
for(t in 1:length(term_x_data)){
  term <- list.files(pattern=paste("terminator_test_", term_x_data[t], sep=''))
  term <- mixedsort(sort(term))
  term_means[t] <- get_limits(term)
}

#Create plots
term_line_plot <- ggplot(term_data, aes(x=strengths, y=values)) + 
  geom_point(stat='identity', color='#009E73', size=5) + 
  geom_line(size=4, stat='identity', color='#009E73') + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        plot.margin = unit(c(2,2,2,2), "cm"), 
        axis.text.x = element_text(size=80), 
        axis.text.y = element_text(size=80), 
        axis.line = element_line(size=1.3), 
        axis.ticks = element_line(size=1.3), 
        axis.ticks.length = unit(0.5, "cm"),
        axis.title.x = element_text(size=80), 
        axis.title.y = element_text(size=80)) +
  xlab("Terminator strength") + 
  ylab("Gene Y protein abundance/\nGene X protein abundance") + 
  panel_border() +
  geom_errorbar(aes(ymin=term_y_data-term_means, ymax=term_y_data+term_means), width=0.05, size=2) +
  geom_vline(xintercept = 0.2, linetype="dashed", color = "dark grey", size=1.3)

#Log scaled plot
term_log_plot <- ggplot(term_data, aes(x=strengths, y=values)) + 
  geom_point(stat='identity', color='#009E73', size=5) + 
  geom_line(size=4, stat='identity', color='#009E73') + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        plot.margin = unit(c(2,2,2,2), "cm"), 
        axis.text.x = element_text(size=80), 
        axis.text.y = element_text(size=80), 
        axis.line = element_line(size=1.3), 
        axis.ticks = element_line(size=1.3), 
        axis.ticks.length = unit(0.5, "cm"),
        axis.title.x = element_text(size=80), 
        axis.title.y = element_text(size=80)) + 
  xlab("Terminator strength (log scale)") + 
  ylab("Gene Y protein abundance/\nGene X protein abundance (log scale)") + 
  panel_border() +
  geom_errorbar(aes(ymin=term_y_data-term_means, ymax=term_y_data+term_means), width=0.05, size=2) +
  scale_y_log10() +
  scale_x_log10() +
  geom_vline(xintercept = 0.2, linetype="dashed", color = "dark grey", size=1.3)

```
Plot RNAse parameters
```{r, echo=FALSE}

setwd('../../data/parameter_testing/rnase/average/')

#List all rnase files
rnase_files <- list.files(pattern="rnase_average_test_")
rnase_files <- mixedsort(sort(rnase_files))

#Set x-axis scale
rnase_x_data <- c(0.0, 0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0)
rnase_y_data <- c()

#Get transcript abundaces at time point 239 seconds and calculate ratio between protein2 and protein1 abundances
for (i in 1:length(rnase_files)){
  p1_data <- read.table(rnase_files[i], header=TRUE) %>% filter(time==239, species=='protein1') %>% select('transcript')
  p2_data <- read.table(rnase_files[i], header=TRUE) %>% filter(time==239, species=='protein2') %>% select('transcript')
 if (nrow(p2_data) != 0){
    rnase_y_data[i] <- p2_data$transcript/p1_data$transcript
  }else{
    rnase_y_data[i] <- 0
  }
}

#Create dataframe needed for ggplot
rnase_data <- data.frame(cbind(rnase_x_data, rnase_y_data))
colnames(rnase_data) <- c('strengths','values')

#Enters working directory where the averages were calculated from
setwd('../')

#Gets averages of each numerical value for error bars
rnase_means <- c()
for(r in 1:length(rnase_x_data)){
  rnase <- list.files(pattern=paste("rnase_test_", rnase_x_data[r], sep=''))
  rnase <- mixedsort(sort(rnase))
  rnase_means[r] <- get_limits(rnase)
}

#Create plots
rnase_line_plot <- ggplot(rnase_data, aes(x=strengths, y=values)) + 
  geom_point(stat='identity', color='#009E73', size=5) + 
  geom_line(size=4, stat='identity', color='#009E73') + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        plot.margin = unit(c(2,2,2,2), "cm"), 
        axis.text.x = element_text(size=80), 
        axis.text.y = element_text(size=80), 
        axis.line = element_line(size=1.3), 
        axis.ticks = element_line(size=1.3), 
        axis.ticks.length = unit(0.5, "cm"),
        axis.title.x = element_text(size=80), 
        axis.title.y = element_text(size=80)) +
  xlab("RNAse strength") + 
  ylab("Gene Y protein abundance/\nGene X protein abundance") + 
  panel_border() +
  geom_errorbar(aes(ymin=rnase_y_data-rnase_means, ymax=rnase_y_data+rnase_means), width=0.05, size=2) +
  geom_vline(xintercept = 5e-3, linetype="dashed", color = "dark grey", size=1.3)

#Log scaled plot
rnase_log_plot <- ggplot(rnase_data, aes(x=strengths, y=values)) + 
  geom_point(stat='identity', color='#009E73', size=5) + 
  geom_line(size=4, stat='identity', color='#009E73') + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        plot.margin = unit(c(2,2,2,2), "cm"), 
        axis.text.x = element_text(size=80), 
        axis.text.y = element_text(size=80), 
        axis.line = element_line(size=1.3), 
        axis.ticks = element_line(size=1.3), 
        axis.ticks.length = unit(0.5, "cm"),
        axis.title.x = element_text(size=80), 
        axis.title.y = element_text(size=80)) +
  xlab("RNAse strength (log scale)") + 
  ylab("Gene Y protein abundance/\nGene X protein abundance (log scale)") + 
  panel_border() +
  geom_errorbar(aes(ymin=rnase_y_data-rnase_means, ymax=rnase_y_data+rnase_means), width=0.05, size=2) +
  scale_y_log10() +
  scale_x_log10() +
  geom_vline(xintercept = 5e-3, linetype="dashed", color = "dark grey", size=1.3)

```
Produce paramters figure
```{r}

setwd("../../data/parameter_testing/figures/")

#Read in genome architecture images
prom_arch <- readPNG("promoter_genome_arch.png")
term_arch <- readPNG("terminator_genome_arch.png")
rnase_arch <- readPNG("rnase_genome_arch.png")

param_figure <- grid.arrange(rasterGrob(prom_arch, vjust=2.5, width=0.9), rasterGrob(term_arch, vjust=2.5, width=0.9), rasterGrob(rnase_arch, vjust=2.5, width=0.9), prom_line_plot, term_line_plot, rnase_line_plot, prom_log_plot, term_log_plot, rnase_log_plot, layout_matrix=rbind(c(1,2,3), c(4,5,6), c(7,8,9)))

ggsave("parameters_figure.png", param_figure, height=70, width=110, limitsize=FALSE, dpi=72)
```
